frontend/test/pytest/test_static_arguments.py

```diff
@@ -123,6 +123,19 @@ def f(x: MyClass, y: int, z: MyClass):
         assert f(MyClass(5), 2, MyClass(5)) == 12
         assert function == f.compiled_function
 
+    def test_default_static_arguments(self):
+        """Test QJIT with static arguments that have a default value."""
+
+        @qjit(static_argnums=[1])
+        def f(y, x=9):
+            if x < 10:
+                return x + y
+            return 42000
+
+        assert f(20) == 29
+        assert f(20, 3) == 23
+        assert f(20, 300000) == 42000
+
```

Comment:
I have two suggestions for this test.

The first one is to use [`pytest.mark.parametrize`](https://docs.pytest.org/en/stable/how-to/parametrize.html#how-to-parametrize-fixtures-and-test-functions). One disadvantage of testing a unit test with multiple inputs is that we won't know which input made it fail. Was it `f(20)`, `f(20, 3)` or `f(20, 300000)`... We can improve on this test by using pytest.mark.parametrize which will essentially create a loop around the inputs and report which specific input broke this test. Like below:

```suggestion
    @pytest.mark.parametrize("args,result", [((20,),29), ((20, 3),23), ((20, 30000),42000)])
    def test_default_static_arguments(self, args, result):
        """Test QJIT with static arguments that have a default value."""

        @qjit(static_argnums=[1])
        def f(y, x=9):
            if x < 10:
                return x + y
            return 42000

        assert(f(*args) == result)

```

However, this suggestion is really unnecessary on light of the next one.

You may not be too familiar with LLVM/MLIR. But when testing code generation, it is better to write a test that verifies the code generated is correct instead of running a unit test. LLVM includes a tool called [`lit`](https://llvm.org/docs/CommandGuide/lit.html) and another one called [FileCheck](https://llvm.org/docs/CommandGuide/FileCheck.html). With it we can write tests that verify that the code generated is what we actually want it to be. For example, that the variable x has a constant value of 9, 3, or 30000 in the code generated. We have lit tests available in [`frontend/test/lit`. ](https://github.com/PennyLaneAI/catalyst/blob/main/frontend/test/lit/test_if_else.py)

Here's an example of what a lit test could look like:

```python
# RUN: %PYTHON -m pytest -s %s | FileCheck %s
# All lit tests have a run command that tells the testing framework what command it will execute.
# This test will run python and specifically the pytest module.
# The -s flag tells pytest to print to stdout. This is very important.
# The %s variable tells lit test that this is the current file.
#
# The output of the python process will be piped to FileCheck.
# The input to FileCheck is both this file and the output of the python process.
# File check will inspect this file and look for check commands.
                                                                               
class TestStaticArguments:                                                     

    @pytest.mark.parametrize("args,result", [((20,),29), ((20, 3),23), ((20, 30000),42000)])
    def test_default_static_arguments(self, args, result):
        """Test QJIT with static arguments that have a default value."""
        
        @qjit(static_argnums=[1])
        def f(y, x=9):
            if x < 10:                    
                return x + y
            return 42000 
            
        f(*args)
        # These are some examples of the check commands.
        # It is saying that there will be a constant variable with value 9
        # One constant variable with value 3
        # And another one with 42000
        # The check commands are executed in order. You may want to add other commands
        # and inspect the output of f.mlir.

        # CHECK: stablehlo.constant dense<9>
        # CHECK: stablehlo.constant dense<3>                         
        # CHECK: stablehlo.constant dense<42000>
        print(f.mlir)
        
if __name__ == "__main__":                   
    pytest.main(["-x", __file__])
```

You may want to modify this template lit test a bit more to make it more readable though. Consider using `# CHECK-LABEL` to find the function that you are interested in.

Reply:
I'm going to resolve this comment since you may not have known about lit tests in the past and it is not a big point of discussion. No need to work on this.